aws/kinesis/core/shard_map.cc

```diff
@@ -63,81 +63,83 @@ boost::optional<uint64_t> ShardMap::shard_id(const uint128_t& hash_key) {
   return boost::none;
 }
 
-void ShardMap::invalidate(TimePoint seen_at) {
-  WriteLock lock(mutex_);
 
+
+void ShardMap::invalidate(const TimePoint& seen_at, const boost::optional<uint64_t> predicted_shard) {
+  WriteLock lock(mutex_);
+  
   if (seen_at > updated_at_ && state_ == READY) {
-    update();
+    if (!predicted_shard || std::binary_search(open_shard_ids_.begin(), open_shard_ids_.end(), *predicted_shard)) {
+      std::chrono::duration<double, std::milli> fp_ms = seen_at - updated_at_;
+      LOG(info) << "Deciding to update shard map for \"" << stream_ 
+                <<"\" with a gap between seen_at and updated_at_ of " << fp_ms.count() << " ms " << "predicted shard: " << predicted_shard;
```

Comment:
Consider logging outside of the lock.

Reply:
Will do.

Reply:
This pattern seemed to exist before. I will leave it as it is for now.